cybench/datasets/torch_dataset.py

```diff
@@ -0,0 +1,207 @@
+from cybench.datasets.dataset import Dataset
+from cybench.util.torch import batch_tensors
+from cybench.util.features import dekad_from_date
+
+
+class TorchDataset(Dataset, torch.utils.data.Dataset):
+    def __init__(
+        self,
+        dataset: Dataset,
+        aggregate_time_series_to: str = None,
+        max_season_window_length: int = None,
+    ):
+        """
+        PyTorch Dataset wrapper for compatibility with torch DataLoader objects
+        :param dataset: Dataset
+        :param aggregate_time_series_to: aggregation resolution for time series data
+                                         ("week" and "dekad" are supported)
+        :param max_season_window_length: maximum length of time series
+        """
+        self._crop = dataset.crop
+        self._df_y = dataset._df_y
+
+        if max_season_window_length is None:
+            self._max_season_window_length = dataset.max_season_window_length
+        else:
+            self._max_season_window_length = max_season_window_length
+
+        self._aggregate_time_series_to = aggregate_time_series_to
+        if aggregate_time_series_to is not None:
+            if aggregate_time_series_to not in ["week", "dekad"]:
+                raise Exception(
+                    f"Unsupported time series aggregation resolution {aggregate_time_series_to}"
+                )
+
+            assert self._max_season_window_length is not None
+            self._dfs_x = {}
+            ts_inputs = [
+                x for x in dataset._dfs_x if "date" in dataset._dfs_x[x].index.names
+            ]
+            for x in dataset._dfs_x:
+                if x not in ts_inputs:
+                    self._dfs_x[x] = dataset._dfs_x[x]
+
+            # combine time series data
+            df_ts = pd.concat(
+                [dataset._dfs_x[x] for x in ts_inputs], join="outer", axis=1
+            )
+
+            # create a dataframe with all dates for every location and year
+            df_all_dates = self._dfs_x[KEY_CROP_SEASON][["cutoff_date"]].copy()
+            df_all_dates["date"] = df_all_dates.apply(
+                lambda r: pd.date_range(
+                    end=r["cutoff_date"],
+                    periods=self._max_season_window_length,
+                    freq="D",
+                ),
+                axis=1,
+            )
+            df_all_dates = df_all_dates.explode("date").drop(columns=["cutoff_date"])
+            df_all_dates.reset_index(inplace=True)
+            df_all_dates.set_index([KEY_LOC, KEY_YEAR, "date"], inplace=True)
+
+            # merge to get all dates in time series data
+            df_ts = df_ts.merge(
+                df_all_dates, how="outer", on=[KEY_LOC, KEY_YEAR, "date"]
+            )
+            del df_all_dates
+
+            # NOTE: interpolate fills data in forward direction.
+            df_ts = df_ts.sort_index().interpolate(method="linear")
+            # fill NAs in the front with 0.0
+            df_ts.fillna(0.0, inplace=True)
+
```

Comment:
I suggest decoupling interpolation from aggregation. Right now using daily data assumes data is complete and equal ts length.

I suggest moving line 50-90 to outside of the aggregation condition.

Reply:
It's also fine if we add this in the future